File: go/test/authorization_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestAuthorizationEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Authorization(nil)
		if ent == nil {
			t.Fatal("expected non-nil AuthorizationEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := authorizationBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "authorization." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_AUTHORIZATION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		authorizationRef01Ent := client.Authorization(nil)
		authorizationRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "authorization"}), "authorization_ref01"))
		authorizationRef01Data["application_id"] = setup.idmap["application01"]

		authorizationRef01DataResult, err := authorizationRef01Ent.Create(authorizationRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		authorizationRef01Data = core.ToMapAny(entityData(authorizationRef01DataResult))
		if authorizationRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if authorizationRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		authorizationRef01DataUp0Up := map[string]any{
			"id": authorizationRef01Data["id"],
		}

		authorizationRef01MarkdefUp0Name := "access_token"
		authorizationRef01MarkdefUp0Value := fmt.Sprintf("Mark01-authorization_ref01_%d", setup.now)
		authorizationRef01DataUp0Up[authorizationRef01MarkdefUp0Name] = authorizationRef01MarkdefUp0Value

		authorizationRef01ResdataUp0Result, err := authorizationRef01Ent.Update(authorizationRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		authorizationRef01ResdataUp0 := core.ToMapAny(entityData(authorizationRef01ResdataUp0Result))
		if authorizationRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if authorizationRef01ResdataUp0["id"] != authorizationRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if authorizationRef01ResdataUp0[authorizationRef01MarkdefUp0Name] != authorizationRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", authorizationRef01MarkdefUp0Name, authorizationRef01ResdataUp0[authorizationRef01MarkdefUp0Name])
		}

	})
}

func authorizationBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "authorization", "AuthorizationTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read authorization test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse authorization test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"authorization01", "authorization02", "authorization03", "application01", "application02", "application03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_AUTHORIZATION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_AUTHORIZATION_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_AUTHORIZATION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/authorization_entity_test.lua
-- Authorization entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("AuthorizationEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Authorization(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = authorization_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "update"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "authorization." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_AUTHORIZATION_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local authorization_ref01_ent = client:Authorization(nil)
    local authorization_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.authorization"), "authorization_ref01"))
    authorization_ref01_data["application_id"] = setup.idmap["application01"]

    local authorization_ref01_data_result, err = authorization_ref01_ent:create(authorization_ref01_data, nil)
    assert.is_nil(err)
    authorization_ref01_data = helpers.to_map(type(authorization_ref01_data_result) == 'table' and authorization_ref01_data_result.data_get and authorization_ref01_data_result:data_get() or authorization_ref01_data_result)
    assert.is_not_nil(authorization_ref01_data)
    assert.is_not_nil(authorization_ref01_data["id"])

    -- UPDATE
    local authorization_ref01_data_up0_up = {
      id = authorization_ref01_data["id"],
    }

    local authorization_ref01_markdef_up0_name = "access_token"
    local authorization_ref01_markdef_up0_value = "Mark01-authorization_ref01_" .. tostring(setup.now)
    authorization_ref01_data_up0_up[authorization_ref01_markdef_up0_name] = authorization_ref01_markdef_up0_value

    local authorization_ref01_resdata_up0_result, err = authorization_ref01_ent:update(authorization_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local authorization_ref01_resdata_up0 = helpers.to_map(type(authorization_ref01_resdata_up0_result) == 'table' and authorization_ref01_resdata_up0_result.data_get and authorization_ref01_resdata_up0_result:data_get() or authorization_ref01_resdata_up0_result)
    assert.is_not_nil(authorization_ref01_resdata_up0)
    assert.are.equal(authorization_ref01_resdata_up0["id"], authorization_ref01_data_up0_up["id"])
    assert.are.equal(authorization_ref01_resdata_up0[authorization_ref01_markdef_up0_name], authorization_ref01_markdef_up0_value)

  end)
end)

function authorization_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/authorization/AuthorizationTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read authorization test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "authorization01", "authorization02", "authorization03", "application01", "application02", "application03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_AUTHORIZATION_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_AUTHORIZATION_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_AUTHORIZATION_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
